File: Roadmap/03 - ESTRUCTURAS DE DATOS/python/CaveroBrandon.py
import sys
"""
* EJERCICIO:
Muestra ejemplos de creación de todas las estructuras soportadas por defecto en tu lenguaje.
Utiliza operaciones de inserción, borrado, actualización y ordenación.
"""


# **** LIST ****
def data_structure_list():
    ordinal_numbers = ['Second', 'Fourth', 'First']  # Create a list
    ordinal_numbers.remove('Fourth')  # Remove an element from the list
    ordinal_numbers.append('Third')  # Append/add an element to the list
    ordinal_numbers.sort()  # Sort the list
    first_element = ordinal_numbers[0]  # Accessing an element within
    ordinal_numbers.clear()  # Clear the list


# **** TUPLE ****
def data_structure_tuple():
    users_category = (1, 2, 'Trial', 'Freemium', 'Premium', 'Pro')  # Create a tuple
    category_slice = users_category[2:]  # Slicing the tuple
    non_users_category = (-1, -2, 'ex member', 'non member')  # Create a second tuple
    all_categories = users_category + non_users_category  # Concatenation of two tuples
    is_premium_in_categories = 'Premium' in all_categories  # Verification that the element 'Premium' is in tuple
    premium_element_count = all_categories.count('Premium')  # Count how many times 'Premium' is in the tuple


# **** DICTIONARIES ****
def data_structure_dictionaries():
    players_stats = {'Jordan': 32292, 'Bryant': 33643, 'James': 46675}  # Create a dictionary
    bryant_points = players_stats['Bryant']  # Accessing the value in the key 'Bryant'
    players_stats['James'] = 46676  # Updating the value of the key 'James'
    players_stats['Kareem'] = 38387  # Adding a new value in the dictionary
    del players_stats['Jordan']  # Removing the element 'Jordan' from the dictionary
    is_jordan_in_stats = 'Jordan' in players_stats  # Checking if the element 'Jordan' is in the dictionary
    dictionary_keys = list(players_stats.keys())  # Getting the keys of the dictionary
    dictionary_values = list(players_stats.values())  # Getting the values of the dictionary
    soccer_stats = {'Messi': 821, 'Ronaldo': 873}  # Create a second dictionary
    players_stats.update(soccer_stats)  # Merge the 'Soccer_stats' into the players stats
    players_stats.clear()  # Clearing the players_stats dictionary


# **** SETS ****
def data_structure_sets():
    students_set = {'Brandon', 'Cavero', 'Brandon', 'Andres'}  # Creating a set
    students_set.add('Moure')  # Adding the element 'Moure' to students_set
    students_set.remove('Cavero')  # Remove the element 'Cavero' from the student_set
    is_andres_in_students = 'Andres' in students_set
    second_students_sets = {'Brandon', 'Kobe', 'Jordan'}  # Create a second set of elements
    union_students_sets = students_set | second_students_sets  # Union of both sets
    intersection_students_sets = students_set & second_students_sets  # Intersection of both sets
    difference_students_sets = students_set - second_students_sets  # Difference of both sets
    symmetric_students_sets = students_set ^ second_students_sets  # Symmetric difference of both sets
    students_set.clear()  # Clearing the students_set

"""
DIFICULTAD EXTRA (opcional) Crea una agenda de contactos por terminal:
- Debes implementar funcionalidades de búsqueda, inserción, actualización y eliminación de contactos.
- Cada contacto debe tener un nombre y un número de teléfono.
- El programa solicita en primer lugar cuál es la operación que se quiere realizar, y a continuación los datos  necesarios para llevarla a cabo.
- El programa no puede dejar introducir números de teléfono no númericos y con más de 11 dígitos (o el número de dígitos que quieras).
- También se debe proponer una operación de finalización del programa. 
"""


def display_menu():
    print('**** MENU ****\n'
          '1. Search contact\n'
          '2. Insert contact\n'
          '3. Update contact\n'
          '4. Remove contact\n'
          '5. Close\n')
    option = input('Select an option: ')
    try:
        option = int(option)
    except ValueError:
        print('Input an integer value between 1 and 5')
        display_menu()

    if 0 < int(option) <= 5:
        if int(option) == 1:
            search_contact()
        elif int(option) == 2:
            insert_contact()
        elif int(option) == 3:
            update_contact()
        elif int(option) == 4:
            remove_contact()
        elif int(option) == 5:
            sys.exit()
    else:
        print('Select a correct option between 1 and 5\n')
        display_menu()


def search_contact():
    print('**** SEARCH CONTACT ****')
    name = input('Input the name of the contact: ')
    if agenda.get(name) is not None:
        print(f'{name}', ':', agenda.get(name))
        display_menu()
    else:
        print(f'No contact with name {name} was found')
        display_menu()


def phone_number_input(name):
    phone_number = input('Phone number: ')
    if phone_number.isnumeric() and len(phone_number) <= 11:
        agenda[name] = int(phone_number)
        print(f'Contact {name} was successfully saved with phone number {phone_number}')
        display_menu()
    else:
        print('Phone number must be a number with less than 11 digits, please try again')
        phone_number_input(name)


def insert_contact():
    print('**** INSERT CONTACT ****')
    name = input('Name of the contact: ')
    phone_number_input(name)


def update_contact():
    print('**** UPDATE CONTACT ****')
    name = input('Enter the contact name to update: ')
    if name in agenda:
        phone_number_input(name)
    else:
        print('Contact was not found, try again')
        update_contact()


def remove_contact():
    print('**** REMOVE CONTACT ****')
    name = input('Enter the contact name to remove: ')
    if name in agenda:
        del agenda[name]
        print(f'Contact {name} was successfully removed from the agenda')
        display_menu()
    else:
        print('Contact was not found, try again')
        remove_contact()


agenda = dict()
display_menu()

File: Roadmap/03 - ESTRUCTURAS DE DATOS/javascript/CaveroBrandon.js
// const { get } = require('http');
const readline = require('readline');

/*
EJERCICIO:
- Muestra ejemplos de creación de todas las estructuras soportadas por defecto en tu lenguaje.
- Utiliza operaciones de inserción, borrado, actualización y ordenación.
*/
// --- List ---
// Population, update operations
let fruitsList = ['Banana', 'Apple'];   // Creating an array
fruitsList[0] = 'Orange';               // Modifying an element based on its position
fruitsList.push('Banana');              // Adding an element to the end
fruitsList.unshift('Banana');           // Adding an element to the beginning
fruitsList.splice(1, 0, 'Strawberry')   // Adding an element on an specific position
fruitsList.splice(4, 1, 'Pear')         // Replacing an element on an specific position
fruitsList.splice(0, 0, 'Grapes', 'Watermelon') // Adding more than one element starting on an specific position
let groceryList = new Array();          // Creating an array
groceryList = fruitsList.concat(['Tomato', 'Pepper'])   // Adding another array to the first array

// Removing operations
groceryList = groceryList.slice(0,3);   // Slicing the array
groceryList.pop();                      // Removing the last element
groceryList.shift();                    // Removing the first element
groceryList.splice(1, 1);               // Removing an element of an specific position
groceryList.reverse();                  // Reversing the array

// Sorting operation
let numberList = [10, 2, 3, 5, 73];
numberList.sort((a, b) => a - b);       // Sorting the array ascending order
numberList.sort((a, b) => b - a);       // Sorting the array descending order

// --- Maps ---
const phoneCharacteristics = new Map();         // Creating a map
phoneCharacteristics.set('Battery', '100');     // Adding an element to the map
phoneCharacteristics.set('Portability', true);  // Adding an element to the map
phoneCharacteristics.set('Wireless', false);    // Adding an element to the map
phoneCharacteristics.set('PhoneNumber', 12321)  // Adding an element to the map
phoneCharacteristics.delete('Battery');         // Remove an element using it's key
let phonePortability = phoneCharacteristics.get('Portability'); // Get an element vßalue using it's key

// --- Set ---
const user = new Set(); // Creating a set
user.add('name');       // Adding an element to the set
user.add('age');        // Adding an element to the set
user.add('weight');     // Adding an element to the set
user.delete('weight');  // Removing an element of the set
user.clear();           // Clear all the set


// --- Objects ---
let newCar = {color: 'red', hasWindows: true, year: 2025};  // Creating an object
newCar.electric = false;    // Adding a new element to the object
newCar.color = 'Blue';      // Updating a value based on a key
delete newCar.hasWindows;   // Removing an element
isColorKeyAvailable = 'color' in newCar // Verifying that an object has a value

/*
DIFICULTAD EXTRA (opcional):
Crea una agenda de contactos por terminal.
- Debes implementar funcionalidades de búsqueda, inserción, actualización y eliminación de contactos.
- Cada contacto debe tener un nombre y un número de teléfono.
- El programa solicita en primer lugar cuál es la operación que se quiere realizar, y a continuación
los datos necesarios para llevarla a cabo.
- El programa no puede dejar introducir números de teléfono no númericos y con más de 11 dígitos.
(o el número de dígitos que quieras)
- También se debe proponer una operación de finalización del programa.
*/
let agenda = [];

const rl = readline.createInterface({
  input: process.stdin,
  output: process.stdout,
});

function getUserInput(questionText) {
    return new Promise((resolve) => {
        rl.question(questionText, (ans) => {
            resolve(ans);
        });
    });
}

async function registerContact() {
    let contact = {};
    const contactName = await getUserInput('Enter the contact name: ');
    const contactPhoneNumber = await getUserInput('Enter contact phone number: ');
    if (isNaN(contact.phoneNumber) || contact.phoneNumber.toString().split('').length > 11) {
        console.log('The entered value must be a number with less than 11 digits');
        showMenu();
    }
    else {
        contact = {name: contactName, phoneNumber: contactPhoneNumber};
        agenda.push(contact);
        console.log('---------------------------------');
        console.log('New entry registered successfully');
        console.log('Contact name: ' + contact.name);
        console.log('Contact phone number: ' + contact.phoneNumber);
        console.log('---------------------------------');
        showMenu();
    }
}

async function updateContact(nameToSearch) {
    const contact = agenda.find(contact => contact.name === nameToSearch);
    if (contact) {
        contact.phoneNumber = await getUserInput('Enter new phone number: ');
        if (isNaN(contact.phoneNumber) || contact.phoneNumber.toString().split('').length > 11) {
            console.log('The entered value must be a number with less than 11 digits');
            showMenu();
        }
        else {
            console.log('---------------------------------');
            console.log('New phone number registered successfully');
            console.log('Contact name: ' + contact.name);
            console.log('Contact phone number: ' + contact.phoneNumber);
            console.log('---------------------------------');
            showMenu();
        }
    }
    else {
        contactNotFoundMessage();
    }
}

async function removeContact(nameToSearch) {
    const contact = agenda.find(contact => contact.name === nameToSearch);
    if (contact) {
        console.log('---------------------------------');
        console.log('The contact was succesfully removed from the agenda');
        console.log('Contact name: ' + contact.name);
        console.log('Contact phone number: ' + contact.phoneNumber);
        const index = agenda.indexOf(contact);
        agenda.splice(index, 1);
        console.log('---------------------------------');
        showMenu();
    }
    else {
        contactNotFoundMessage();
    }
}

async function getPhoneNumberByName(nameToSearch) {
    // Find the contact with the matching name in the global 'agenda' list
    const contact = agenda.find(contact => contact.name === nameToSearch);
    if (contact) {
        console.log('---------------------------------');
        console.log('Contact "' + nameToSearch + '" found in the agenda');
        console.log(nameToSearch + ': ' + contact.phoneNumber);
        console.log('---------------------------------');
        showMenu()
    }
    else {
        contactNotFoundMessage();
    }
}

function isAgendaEmpty() {
    if (agenda.length === 0) {
        console.log('---------------------------------');
        console.log('Agenda is empty, try adding some contacts first');
        console.log('---------------------------------');
        showMenu();
    }
}

function contactNotFoundMessage() {
    console.log('---------------------------------');
    console.log('The contact was not found in the agenda');
    console.log('---------------------------------');
    showMenu();
}

async function showMenu() {
    console.log('Select an option');
    console.log('1. Register a new contact (Insert)');
    console.log('2. Update an existing contact');
    console.log('3. Remove an existing contact');
    console.log('4. Search a contact')
    console.log('5. Exit');
    const option = await getUserInput('Select your option (1-5): ');
    switch (option) {
        case '1':
            registerContact();
            break;
        case '2':
            isAgendaEmpty();
            const nameToUpdate = await getUserInput('Enter the contact name to update: ');
            updateContact(nameToUpdate);    
            break;
        case '3':
            isAgendaEmpty();
            const nameToRemove = await getUserInput('Enter the contact name to remove: ');
            removeContact(nameToRemove);
            break;
        case '4':
            isAgendaEmpty();
            const nameToSearch = await getUserInput('Enter the contact name to search: ');
            getPhoneNumberByName(nameToSearch);
            break;
        case '5':
            process.exit(0);
        default:
            console.log('---------------------------------');
            console.log('Invalid option. Please select a valid option (1-5)');
            console.log('---------------------------------');
            showMenu();
    }
}

showMenu();
